### PR TITLE
Allow compound expressions to begin with "not"

### PR DESCRIPTION
**Pull request by plastikos**
This eliminates the necessity of the following construct:

```
>$ salt -C "* and not foo" test.ping
```

and allows a more natural grammar:

```
>$ salt -C "not foo" test.ping
```

This is mildly tested.  I tried running the unit tests but they seem to be broken since the merge of tornado.  I have run the unit tests on `2015.2`.

### CODE REVIEW

**Comment by plastikos** [comment]
There are two more patches built on top of this:

    * Specifying the delimiter character in complex matches
    * Specifying complex matches as a list rather than a string (allows spaces to be used in the matches).

I'll submit those once this is accepted.

**Comment by jfindlay** [comment]
@plastikos there seems to be a test failure.

**Comment by jfindlay** [comment]
Actually not, this test is failing on other pull requests.

**Comment by plastikos** [comment]
Does anyone else have problems with the test suite starting on the {{develop}} branch?  The test suite won't progress past {{Waiting for minions(minion, sub_minion) to connect back}}.

**Comment by jfindlay** [comment]
@plastikos, you can see the debug log from the test suite at `/tmp/salt-runtests.log`.  The only time I've seen that is when stale processes from previous tests runs linger around holding onto zmq sockets or something.

**Comment by plastikos** [comment]
So now I'm working on fixing the unit tests in order to have confidence to merge this and two new features I have.

See https://github.com/saltstack/salt/pull/22992
There will be more fixes as I dig into the broken unit tests.

**Comment by plastikos** [comment]
Okay - I have now fixed my test start-up problems.  It turned out that I needed to update to tornado 4.0.  I have no concerns about this being merged.

File: doc/topics/targeting/compound.rst [hunk 1]
```diff
@@ -41,18 +41,11 @@ That same example expressed in a :term:`top file` looks like the following:
...
-
-.. code-block:: bash
-
-    salt -C '* and not G@kernel:Darwin' test.ping
-
 Excluding a minion based on its ID is also possible:
 
 .. code-block:: bash
 
-    salt -C '* and not web-dc1-srv' test.ping
```
Comment (thatch45) [inline doc/topics/targeting/compound.rst]: Maybe add a versionadded directive in here?

**Pull request merged**